### PR TITLE
feat: support workflow multi version recovery

### PR DESCRIPTION
**Pull request by luckylhb90**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Feat:  support workflow multi version recovery

image

image

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A
- [ ] Test B

### CODE REVIEW

**Review by RahulVadisetty91 — COMMENTED** [review]
w5678u

Refactor to catch more specific exceptions that are likely to occur in your service.

**Comment by warren830** [comment]
@luckylhb90 @RahulVadisetty91 @rainchen @hobo-l-20230331 Hi guys, any update for this pr? We really really need this!

**Comment by laipz8200** [comment]
Hi @luckylhb90, sorry for the late reply, would you like to resolve those conflicts?

**Comment by warren830** [comment]
@crazywoola hello, need to fix python style check

**Review by crazywoola — APPROVED** [review]
LGTM @laipz8200

**Comment by laipz8200** [comment]
Hi @luckylhb90, could you sync the code with the main branch? We've added a new Mypy checker and it needs to be passed.

File: api/services/workflow_service.py [hunk 1]
```diff
@@ -76,6 +78,38 @@ def get_published_workflow(self, app_model: App) -> Optional[Workflow]:
...
+
+        if len(workflows) > 1:
+            workflows[1].version = "latest"
+
+        for workflow in workflows:
+            try:
+                # Try to parse the version as a datetime object
+                version_datetime = datetime.strptime(workflow.version, "%Y-%m-%d %H:%M:%S.%f")
+                # If successful, reformat it to the desired format
+                workflow.version = version_datetime.strftime("%Y-%m-%d %H:%M:%S")
```
Comment (laipz8200) [inline api/services/workflow_service.py:106]: update a db model for API response it's not a good way, please consider doing this convert in the front end.

The same issue is on L99.

File: api/controllers/console/app/workflow.py [hunk 2]
```diff
@@ -454,6 +473,7 @@ def get(self, app_model: App):
     WorkflowDraftRunIterationNodeApi, "/apps/<uuid:app_id>/workflows/draft/iteration/nodes/<string:node_id>/run"
 )
 api.add_resource(PublishedWorkflowApi, "/apps/<uuid:app_id>/workflows/publish")
+api.add_resource(PublishedAllWorkflowApi, "/apps/<uuid:app_id>/workflows/publish/all")
```
Comment (laipz8200) [inline api/controllers/console/app/workflow.py:476]: IMO, this route name should be `/apps/<uuid:app_id>/workflows`

**Review by laipz8200 — CHANGES_REQUESTED** [review]

**Comment by laipz8200** [comment]
Since the author hasn’t responded for a long time, I think we can close this PR and complete it in #11990.

**Pull request closed without merging**